lib/msfenv.rb

```diff
@@ -41,3 +41,42 @@ def noop_error_formatter.message_for(_spot)
 end
 
 MsfAutoload.instance
+
+# @deprecated
+def print_line(msg)
+  warn '[DEPRECATION] The global print_line is deprecated.'
+  warn "Called from: #{caller(0, 2)}"
```

Comment:
If we put this on one line, it'll be easier to grep for - especially in multi-threaded environments


```ruby
def _warn_deprecation_message(method)
  stack_size = 3
  warning_message = "[DEPRECATION] The global method #{method.inspect} is deprecated, please raise a Github issue with this output. Called from: #{caller(1, stack_size).to_a}"
  warn(warning_message)
  # Additionally write to ~/.msf4/logs/framework.log - as this gets attached to Github issues etc
  elog(warning_message)
end

# @deprecated In most scenarios you should delegate to either a framework module object, or Rex::Ui::Text::DispatcherShell etc
def print_line(msg)
  _warn_deprecation_message __method__
  $stdout.puts(msg)
end

# @deprecated In most scenarios you should delegate to either a framework module object, or Rex::Ui::Text::DispatcherShell etc
def print_warning(msg)
  _warn_deprecation_message __method__
  $stderr.puts(msg)
end

# @deprecated In most scenarios you should delegate to either a framework module object, or Rex::Ui::Text::DispatcherShell etc
def print_good(msg)
  _warn_deprecation_message __method__
  $stdout.puts(msg)
end

# @deprecated In most scenarios you should delegate to either a framework module object, or Rex::Ui::Text::DispatcherShell etc
def print_error(msg, exception = nil)
  _warn_deprecation_message __method__

  unless exception.nil?
    msg += "\n    Call Stack:"
    exception.backtrace.each {|line|
      msg += "\n"
      msg += "\t #{line}"
    }
  end

  $stderr.puts(msg)
end
```